receiver/otlpreceiver/otlphttp.go

```diff
@@ -38,39 +38,59 @@ func (*xProtobufMarshaler) ContentType() string {
 
 var jsonMarshaler = &jsonpb.Marshaler{}
 
-// errorHandler encodes the HTTP error message inside a rpc.Status message as required
-// by the OTLP protocol.
-func errorHandler(w http.ResponseWriter, r *http.Request, errMsg string, statusCode int) {
-	var (
-		msg []byte
-		s   *status.Status
-		err error
-	)
-	// Pre-computed status with code=Internal to be used in case of a marshaling error.
-	fallbackMsg := []byte(`{"code": 13, "message": "failed to marshal error message"}`)
-	fallbackContentType := "application/json"
+// writeError encodes the HTTP error inside a rpc.Status message as required by the OTLP protocol.
+func writeError(w http.ResponseWriter, contentType string, err error, statusCode int) {
+	s, ok := status.FromError(err)
+	if ok {
+		writeResponse(w, contentType, statusCode, s.Proto())
+	} else {
+		writeErrorMsg(w, contentType, err.Error(), statusCode)
+	}
+}
 
+// writeErrorMsg encodes the HTTP error message inside a rpc.Status message as required
+// by the OTLP protocol.
+func writeErrorMsg(w http.ResponseWriter, contentType string, errMsg string, statusCode int) {
+	var s *status.Status
 	if statusCode == http.StatusBadRequest {
 		s = status.New(codes.InvalidArgument, errMsg)
 	} else {
-		s = status.New(codes.Internal, errMsg)
+		s = status.New(codes.Unknown, errMsg)
 	}
 
-	contentType := r.Header.Get("Content-Type")
+	writeResponse(w, contentType, statusCode, s.Proto())
+}
+
+// errorHandler encodes the HTTP error message inside a rpc.Status message as required
+// by the OTLP protocol.
+func errorHandler(w http.ResponseWriter, r *http.Request, errMsg string, statusCode int) {
+	writeErrorMsg(w, r.Header.Get("Content-Type"), errMsg, statusCode)
+}
+
+// Pre-computed status with code=Internal to be used in case of a marshaling error.
+var fallbackMsg = []byte(`{"code": 13, "message": "failed to marshal error message"}`)
```

Comment:
Give a more self representative name such as marshalingErrorMsg = = []byte(`{"code": 13, "message": "failed to marshal error message"}`)